source/common/upstream/outlier_detection_impl.cc

```diff
@@ -309,10 +309,11 @@ void DetectorImpl::initialize(Cluster& cluster) {
   }
 
   if (config_.successfulActiveHealthCheckUnejectHost() && cluster.healthChecker() != nullptr) {
-    cluster.healthChecker()->addHostCheckCompleteCb([this](HostSharedPtr host, HealthTransition) {
+    cluster.healthChecker()->addHostCheckCompleteCb([this](HostSharedPtr host, HealthTransition, HealthState state) {
       // If the host is ejected by outlier detection and active health check succeeds,
       // we should treat this host as healthy.
-      if (!host->healthFlagGet(Host::HealthFlag::FAILED_ACTIVE_HC) &&
+      if (state == HealthState::Healthy &&
```

Comment:
This would delay the uneject based on the heath check configuration (interval, healthy threshold etc).  This would essentially render the feature useless - by the time, outlier would have unejected because elapsed time is passed

Reply:
??? The feature is useless today: It unejects the host at the next hc run after the eject even if the result was failure but we still didn't reach the unhealthy threshold.
The feature is named: **successful**_active_health_check_uneject_host. For me this means **only** successful hc can uneject.
Please explain why good to uneject for failed hc. I abandon the pr and the issue in that second.

Reply:
I do not think it will uneject if the health check fails. It should only uneject if atleast one health check succeeds. Are you seeing failed health check unejecting? Atleast the tests I have added when I implemented this ensures the uneject happens on first succesful healthcheck.

Reply:
Yes, I see failed health check unejecting. See the referenced issue for details.
Right, your test is there and ensures the uneject happens on first succesful healthcheck, but there is nothing about failed healthcheck. It is not even possible make such today since callback doesn't get this information.
If unhealthy_threshold is 3 let's say, this hc sequence can happen:

success, !FAILED_ACTIVE_HC
fail, !FAILED_ACTIVE_HC
fail, !FAILED_ACTIVE_HC
fail, FAILED_ACTIVE_HC
fail, FAILED_ACTIVE_HC
...

Callback just checks !FAILED_ACTIVE_HC. If the ejection happens after the 1st or 2nd hc run then next is fail but will uneject.
FAILED_ACTIVE_HC is the host's health flag and in general it is not equal with the current hc run result. Do I misunderstand something?

I tested this with 1.28.0, but I can try double check with the latest. Also I'm using thrift health check but I think that shouldn't matter.

Reply:
But if you check `healthy && **!host->healthFlagGet(Host::HealthFlag::FAILED_ACTIVE_HC**` then you will only uneject on host transitioning to healthy via active health checks(which might be 3 or 10). - That was my point. But active health checks are configured in such a way to the threshold will finish before ejection time - it may fine otherwise this feature wont be of that value.,

Reply:
Yes.  #3 was the original intention. Now that I think of it more, it is OK to honour health check configuration and uneject only if FAILED_ACTIVE_HC is not set to align with host health check flag. But we need to document that when this feature is used, it will only be useful if health check is configured in a way that total time for successful health check is less than the ejection time.

Hope this makes sense. @wbpcode  @zuercher  WDYT?